### PR TITLE
WIP: Centre note support

### PR DESCRIPTION
**Pull request by nikclayton**
This is most of the way to supporting centre notes. "Most of the way" because only the single number input mode has been modified, to allow for feedback.

Key UI changes:

- Three buttons for input type, instead of the number/note toggle button of before. I've re-arranged the input buttons in to a 3x3 grid for the numbers, a 3x1 column for the input modes, and a 2x1 column for the other controls.

- Corner notes numbers are not fixed in position in a 3x3 grid in the cell. Instead, they're in the 2x3 grid, allowing for 6 corner notes maximum (to allow room for the centre note)

### CODE REVIEW

**Comment by nikclayton** [comment]
Here's a screenshot of the UI with this PR. The first three rows all contain cells with centre notes, the remaining notes are corner notes.

Screenshot_1630516758

**Comment by ogarcia** [comment]
Hi @nikclayton 

The idea of rearranging the input buttons in a 3x3 grid is great. But that you have two methods to add notes, not so much. I have a lot of feedback of users and don't like these changes.

I think that is better have a grid of 4x3 buttons. 3x3 numeric pad and one column with delete, change entry mode and toggle notes/numbers. Something like the mockups of #115 

But in your screenshot, I like the icons. They can be used in toggle button to switch between entry notes or numbers.

**Comment by ogarcia** [comment]
When I say "have two methods to add notes" I refer to keyboard, but it can be a config option. Use corner notes or center notes.

**Comment by nikclayton** [comment]
Thanks for the feedback.

I think we disagree about the "or" in:

> Use corner notes or center notes.

I don't think that matches the feature request, or how this is intended to be used.

Corner ("edge" might be a better term) marks indicate "This number might be in this cell, but so might other numbers".

Centre marks indicate "This cell can *only* be one of these candidates".

And it's realistic/natural to use both while solving a puzzle. 

For an example of this in action, see e.g. https://www.youtube.com/watch?v=9m9t8ie9-EE (1m50 to 4m40 in that video explains the technique, and when to use corner marks and when to use centre marks, when solving the puzzle both are used). https://cracking-the-cryptic.web.app/sudoku/PrfmJPgdft is the web app with the same functionality.

I agree with you that the UI changes in #115 also make sense, but if I tried to make those changes here too the PR would be even larger.

Speaking of UI changes, do you have a specific set of devices (emulators?) that you use when checking that new layouts work OK?

File: app/src/main/java/org/moire/opensudoku/gui/SudokuBoardView.java [hunk 1]
```diff
@@ -547,15 +551,26 @@ protected void onDraw(Canvas canvas) {
...
+                        if (!cell.getCornerNote().isEmpty()) {
+                            Collection<Integer> numbers = cell.getCornerNote().getNotedNumbers();
+                            int i = 0;
                             for (Integer number : numbers) {
-                                int n = number - 1;
-                                int c = n % 3;
-                                int r = n / 3;
-                                canvas.drawText(Integer.toString(number), cellLeft + c * noteWidth + 2, cellTop + mNoteTop - noteAscent + r * noteWidth - 1, mCellNotePaint);
+                                int c = i % 3;
+                                // First 3 numbers draw on row 1, remaining numbers draw on row 2.
```
Comment (nikclayton) [inline app/src/main/java/org/moire/opensudoku/gui/SudokuBoardView.java]:
```suggestion
                                // First 3 numbers draw on row 0, remaining numbers draw on row 2.
```

**Comment by ogarcia** [comment]
I'm migrating Open Sudoku development from GitHub to GitLab. To make the migration as clean as possible I'm going to close this PR. If you are still interested in it, please open it again in [GitLab](https://gitlab.com/opensudoku/opensudoku) following the [contributing guide](https://gitlab.com/opensudoku/opensudoku/-/blob/develop/CONTRIBUTING.md).

Thank you very much for your understanding.

**Pull request closed without merging**